### PR TITLE
Challenge proposal: NonNull

### PR DESCRIPTION
**Pull request by tautschnig**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by zhassan-aws — COMMENTED** [review]
Thanks @tautschnig. It seems to me that memory safety for the most of the listed functions would directly follow from their pre-conditions. Thus, I suggest we add the verification of at least one of the modules that use `NonNull` to this challenge, so that we can verify that the pre-conditions are not too overconstraining. One such module is [`Unique`](https://docs.rs/unique/latest/unique/struct.Unique.html).

What do you think?

**Comment by tautschnig** [comment]
> Thanks @tautschnig. It seems to me that memory safety for the most of the listed functions would directly follow from their pre-conditions. Thus, I suggest we add the verification of at least one of the modules that use `NonNull` to this challenge, so that we can verify that the pre-conditions are not too overconstraining. One such module is [`Unique`](https://docs.rs/unique/latest/unique/struct.Unique.html).
> 
> What do you think?

I get your concern - but then contracts (and harnesses) for `Unique` are already PR'ed in https://github.com/model-checking/verify-rust-std/pull/45 in flight, so that's perhaps already sorted?

File: doc/src/challenges/0006-nonnull.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+1. `NonNull<T>::slice_from_raw_parts`
+1. `NonNull<T>::sub_ptr`
+1. `NonNull<T>::sub`
+1. `NonNull<T>::swap`
+1. `NonNull<T>::to_raw_parts`
+1. `NonNull<T>::with_addr`
+1. `NonNull<T>::write_bytes`
+1. `NonNull<T>::write_unaligned`
+1. `NonNull<T>::write_volatile`
+1. `NonNull<T>::write`
```
Comment (feliperodri) [inline doc/src/challenges/0006-nonnull.md:70]: could you fix the numbering?

**Review by feliperodri — APPROVED** [review]

File: doc/src/challenges/0006-nonnull.md (hunk 1 above)
Comment (tautschnig) [reply doc/src/challenges/0006-nonnull.md]: Sure, filtered through `nl -s.`. As markdown, however, it would render correctly either way.

**Pull request merged**